renderdoc/driver/metal/metal_device.cpp

```diff
@@ -23,12 +23,25 @@
  ******************************************************************************/
 
 #include "metal_device.h"
+#include "metal_helpers_bridge.h"
+#include "metal_library.h"
+#include "metal_manager.h"
 
 WrappedMTLDevice::WrappedMTLDevice(MTL::Device *realMTLDevice, ResourceId objId)
     : WrappedMTLObject(realMTLDevice, objId, this, GetStateRef())
 {
   wrappedObjC = AllocateObjCWrapper(this);
+  Construct();
+  GetResourceManager()->AddCurrentResource(objId, this);
+}
+
+void WrappedMTLDevice::Construct()
```

Comment:
This is a strange pattern, why split out some of the constructor into this function? is there going to be another constructor in future, and if so why?

Reply:
It is used in the future for the replay ie in

```
ReplayStatus Metal_CreateReplayDevice(RDCFile *rdc, const ReplayOptions &opts, IReplayDriver **driver)
{
...
  WrappedMTLDevice *wrappedMTLDevice = new WrappedMTLDevice();

  MetalReplay *replay = wrappedMTLDevice->GetReplay();
  replay->SetProxy(isProxy);
  ReplayStatus status = wrappedMTLDevice->Initialise(initParams, ver, opts);
...
}
```
and the `Initialise` method creates the `real` `MTLDevice`

That could be re-organized to do the `MTLDevice` creation in `Metal_CreateReplayDevice` and pass it to `WrappedMTLDevice::WrappedMTLDevice(MTL::Device *realMTLDevice, ResourceId objId)` and the remove the requirement for `WrappedMTLDevice::WrappedMTLDevice()` and `void WrappedMTLDevice::Construct()` ?

Reply:
Yes the idea of the `InitParams` is that this struct can be used for any parameters 'outside' the normal flow of serialised API calls, those on the initial function or calls used to bootstrap the API in the first place. It is then serialised without any of the wrappers inside the `CreateReplayDevice` function, which uses the parameters to initialise the API and it would then pass the created metal device into the `WrappedMTLDevice` constructor.

You could organise it such that the `WrappedMTLDevice` creates itself in `Initialise()` which is what vulkan does, but in that case you should change it so the constructor never takes any parameters and the function call to create a device is a member function. It doesn't really fit with the more OO design of metal and how it organises its functions, so I'd suggest sticking closer to how D3D11/D3D12 organise as above.

In practice it doesn't look like metal has many parameters for device creation? I don't see any way of specifying a version etc so I assume it just always works at the newest version. It does look like you might want at minimum some kind of device identifier so you can pick the matching device in any system with multiple GPUs for `MTLCopyAllDevices` (though I don't know if in practice there exist any apple devices that have multiple GPUs so that might be an entirely theoretical concept?). This may become kind of a trivial distinction.

Reply:
1f9a217577fe36c8bc218da0bbb2f3e721f22011